sdk/core/core/inc/az_json.h

```diff
@@ -182,81 +196,233 @@ AZ_NODISCARD az_result az_json_token_get_string(az_json_token const* token, az_s
+// TODO: Consider adding char* overloads to make passing string literals as property names easier.
+/**
+ * @brief Appends the UTF-8 property name (as a JSON string) which is the first part of a name/value
+ * pair of a JSON object.
  *
- * @param json_builder A pointer to an az_json_builder instance containing the buffer to append the
- * object to.
- * @param name A span containing the JSON object's name.
- * @param token A pointer to the az_json_token to append as the object's value.
- * @return AZ_OK if the token was appended successfully.<br>
- * AZ_ERROR_BUFFER_OVERFLOW if the buffer is too small.
+ * @param[in] json_builder A pointer to an #az_json_builder instance containing the buffer to append
+ * the property name to.
+ * @param[in] name The UTF-8 encoded property name of the JSON value to be written. The name is
+ * escaped before writing.
+ *
+ * @return An #az_result value indicating the result of the operation:
+ *         - #AZ_OK if the property name was appended successfully
+ *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
  */
 AZ_NODISCARD az_result
-az_json_builder_append_object(az_json_builder* json_builder, az_span name, az_json_token token);
+az_json_builder_append_property_name(az_json_builder* json_builder, az_span name);
 
-/*
- * @brief az_json_builder_append_array_item appends an array item to the JSON buffer.
+/**
+ * @brief Appends a boolean value (as a JSON literal `true` or `false`).
  *
- * @param json_builder A pointer to an az_json_builder instance containing the buffer to append the
- * array item to.
- * @param token A pointer to the az_json_token representing the array item.
- * @return AZ_OK if the token was appended successfully.<br>
- * AZ_ERROR_BUFFER_OVERFLOW if the buffer is too small.
+ * @param[in] json_builder A pointer to an #az_json_builder instance containing the buffer to append
+ * the boolean to.
+ * @param[in] value The value to be written as a JSON literal true or false.
+ *
+ * @return An #az_result value indicating the result of the operation:
+ *         - #AZ_OK if the boolean was appended successfully
+ *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
+ */
+AZ_NODISCARD az_result az_json_builder_append_bool(az_json_builder* json_builder, bool value);
+
+// TODO: Consider removing for now until we have complete double formatting and parsing.
+/**
+ * @brief Appends a double number value.
+ *
+ * @param[in] json_builder A pointer to an #az_json_builder instance containing the buffer to append
+ * the number to.
+ * @param[in] value The value to be written as a JSON number.
+ *
+ * @return An #az_result value indicating the result of the operation:
+ *         - #AZ_OK if the number was appended successfully
+ *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
+ */
+AZ_NODISCARD az_result az_json_builder_append_number(az_json_builder* json_builder, double value);
```

Comment:
Why did we prefer having more functions to append than just one that uses the polymorphic az_json_token structure?
The benefit of having a Union type for the az_json_token that can represent any token is useless if there is a function that takes each type of the Union instead.
Think of the Union type az_json_token as an abstract class in OOP, and then we have multiple concrete clases from it.
Then we have on function to use each concrete class instead of just one method that receives the abstract class.

Reply:
We may want to have that in addition, if there is a specific need, but there are a few cons of only having such an API:
1) API usability for the simple/common case is worst.
2) Requires going through a switch unnecessarily for each token which is bad for perf.

Reply:
> This means az_json_token to be usless and it can be removed,

Yes, that is quite possible. Once the parse side is in place, I think I will remove `az_json_token`.

> See example above, working with az_json_token in both directions (appending and parsing) gives an easy api to customers, than having multiple functions.

Yes, I understand. I think just the `az_json_token_kind` enum is sufficient for that use case. We don't want to process all the token values unnecessarily when parsing, if the caller just wants to skip past them.

> but as soon as they need to query something from netwok that returns az_span, then it's not a good API

That will hopefully be solved by the upcoming parser changes. If it isn't addressed then, please definitely bring it up then and we can discuss it further :)

Reply:
oh, `az_json_parse_by_pointer` is probably my favorite API!  Make parsing super easy, hopefully we can add something similar